chunked_stream/lib/src/chunked_stream_iterator.dart

```diff
@@ -23,197 +23,179 @@ import 'dart:math';
 /// `List<int>` for each event.
 ///
 /// Note. methods on this class may not be called concurrently.
-class ChunkedStreamIterator<T> {
-  StreamSubscription<List<T>> _subscription;
-  bool _isDone = false;
-  List<T> _lastChunk;
-  Object _error;
-  StackTrace _stackTrace;
-  void Function() _wakeup;
-
-  List<T> _buffered = [];
-
-  ChunkedStreamIterator._(Stream<List<T>> stream) {
-    ArgumentError.checkNotNull(stream, 'stream');
-
-    _subscription = stream.listen((data) {
-      _lastChunk = data;
-      _wakeup();
-    }, onDone: () {
-      _isDone = true;
-      _wakeup();
-    }, onError: (e, st) {
-      _error = e;
-      _stackTrace = st;
-      _wakeup();
-    });
-    _subscription.pause();
+abstract class ChunkedStreamIterator<T> {
+  factory ChunkedStreamIterator(Stream<List<T>> stream) {
+    return _ChunkedStreamIterator<T>(stream);
   }
 
-  /// Create a [ChunkedStreamIterator] over [stream].
-  factory ChunkedStreamIterator(Stream<List<T>> stream) =>
-      ChunkedStreamIterator._(stream);
+  /// Returns a list of the next [size] elements in original chunks. The size of
+  /// the chunks are not guaranteed to be consistent.
+  ///
+  /// Returns a list with a total of less than [size] elements split in chunks
+  /// if the end of stream is encounted before [size] elements are read.
+  ///
+  /// If an error is encountered before reading [size] elements, the error
+  /// will be thrown.
+  Future<List<List<T>>> readPreservingChunks(int size);
```

Comment:
What is the motivation for this method? Do we need to expose it?

What is the behavior if I ask for `size = 7` and stream contains chunks of size 4? I'm not sure I can read that from the documentation.

Reply:
hm I think you're right.

I started writing this method because I thought there might be instances where users of the library might know the total number of elements to receive in advance, but wish to preserve the chunks - say the original stream was matrix data, or a stream of input/parameters/output and the program was doing some form of online processing. I just took a look at the areas where I used this function in the tar code, and it turns out I only ended up using this function to skip through the stream and avoid the extra work involved in `read` aha.

I'll remove this function for now, and with this gone we should have feature parity with the previous version of `ChunkedStreamIterator`, and if it turns out that this might be helpful, it shouldn't be too hard to add it back in.